### ISSUE TITLE
Use HTTP2 CONTINUATION frame when sending large response headers

### ISSUE THREAD

**Issue by muodov** [haproxy/haproxy#249]
## Output of `haproxy -vv` and `uname -a`

```
# /usr/sbin/haproxy -vv
HA-Proxy version 2.0.5-1~bpo10+1 2019/08/22 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.0.5=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-format-truncation -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1c  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.1c  28 May 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```
```
# uname -a
Linux debiantest 4.19.0-5-amd64 #1 SMP Debian 4.19.37-5+deb10u2 (2019-08-08) x86_64 GNU/Linux
```

## What's the configuration?

```
global
	log /dev/log	local0 debug
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
	stats timeout 30s
	tune.bufsize 32768
	user haproxy
	group haproxy
	daemon

	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private

	# See: https://ssl-config.mozilla.org/#server=haproxy&server-version=2.0.3&config=intermediate
        ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
        ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

defaults
	log	global
	mode	http
	option	httplog
	#option	dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http

frontend fe-https
    bind :::443 ssl crt /certs_public/example.com.pem v4v6 alpn h2,http/1.1
    mode http
    option httplog
    option forwardfor header surfly-forwarded
    option http-server-close

    use_backend test_app

backend test_app
    # the app sends a response with a lot of cookies (real-life scenario), so the resulting HEADERS frame becomes larger than spec default (16384 bytes)
    server 127.0.0.1:5000 127.0.0.1:5000
```

## Steps to reproduce the behavior

1. set up a simple backend that sets a lot of cookies in a single response. The total size of cookie data has to be more than 16k
2. Set `tune.bufsize` to a higher value, so haproxy allows that large response
3. Use Chrome to open the page

## Actual behavior

haproxy translates the response to a single h2 HEADERS frame. The size exceeds the negotiated SETTINGS_MAX_FRAME_SIZE value (which is 16384 by default). As a result, Chrome rejects the response with FRAME_SIZE_ERROR.

## Expected behavior

As per HTTP2 spec, HEADERS frame should be split into several CONTINUATION frames. See https://http2.github.io/http2-spec/#CONTINUATION

## Do you have any idea what may have caused this?

As I understand, haproxy can already _decode_ CONTINUATION frames in requests and backend responses. However, it cannot generate those frames when it is needed.

## Do you have an idea how to solve the issue?

I'm not familiar with haproxy architecture, but the general logic should be to keep track of the current MAX_FRAME_SIZE (it is negotiated in SETTINGS frame), and split the generated data frames accordingly

**Comment by wtarreau** [comment]
Simple question, is your use case real or just a test ? I mean, I've been really hoping *not* to implement CONTINUATION on output because this adds a lot more mess in the multiplexer, as you are not allowed to mix other frames from other streams in between, and doing it without introducing a trivial DoS nor introducing new classes of bugs or performance degradation is quite challenging. We indeed implemented decoding during 1.9 and backported it to 1.8 because one client was abusing it by sending very small frames. But as long as we can avoid this mess on output, we'll feel better :-/  That's why I'm interested in your honest feedback on this one, and if possible the situation where you need more than 16kB of headers over H2 so that I may possibly take some shortcuts.

Note, I'm turning this to a feature request.

**Comment by muodov** [comment]
@wtarreau I appreciate your quick response, thanks!

For us this case is real. We're building a content-rewriting proxy at Surfly, so you can browse any website through our proxy on a different domain. One side-effect here is that we need to transfer cookies from the real domain to the "proxied" domain.
We have a real customer that is storing quite a lot of data in cookies. On the original site, those cookies are set in several different responses, but during transfer, we are setting them all at once in a single response.

I can imagine it could complicate things on haproxy side... But from our experience, large numbers of cookies are not uncommon, especially in enterprise applications. We had to increase the backend limits quite a lot.

**Comment by wtarreau** [comment]
Yep I see. I've also already met very large cookies in enterprises, it just turns out that till now H2 and large cookies were mutually exclusive: people use H2 for speed and they have large cookies by erratic architecture choices. And feeding extremely large cookies to the client means these cookies are sent with all requests. Worse, they may often induce a flush of the HPACK headers table resulting in very bad compression. It's really easy to multiply the client's upload bandwidth by 100 or 1000 when using large cookies, sometimes the upstream bandwidth is even higher than the downstream one!

So I think that it will have to be added to the todo list anyway and that there won't be many options to cheat around :-/ We can *try* to add this to 2.1 but I cannot make any promise. Thanks for the explanation.

**Comment by chipitsine** [comment]
sometimes, openid connect servers set huge cookies, I seen 16kb is not enough

**Comment by chipitsine** [comment]
also, grafana app could use huge url sometimes

**Comment by wtarreau** [comment]
Hi!

I implemented this right now in 2.1-dev. I found a clean way to do it which doesn't affect the common code path and which is able to fragment on any arbitrary byte, thus it does support individual headers >16kB or URLs larger than 16kB. It's in commit cb985a4. I *think* it could be backported once validated in 2.1, as the changes are fairly isolated, though I'm not sure yet about what it will do for legacy mode (as opposed to HTX).

**Comment by muodov** [comment]
@wtarreau this is awesome! It would really help us if it's backported (we're still running old Debian atm), but in any case, thank you for addressing this! 👍

**Comment by wtarreau** [comment]
I may possibly consider backporting it. I'm just hesitant because I know that having support for certain features in certain sub-versions causes trouble on issue reports for people running on their distro vendor's version. I don't think 2.0 is officially shipped in any distro yet, so if everyone still relies either on official sources or on Vincent's up-to-date packages, then we can avoid this fragmentation and it can make sense to backport it. Please do me a favor, have a look around to see if any distro already ships 2.0, that will help me :-)

**Comment by muodov** [comment]
We are still running jessie, so we're stuck with 1.8 for now 😅 
I understand that there were significant changes since then and it can be tricky to backport.
If it's not possible to backport to 1.8, this will be a problem for us until we upgrade the stack (which we should do anyways, but it takes time because of on-premise clients).
If someone is already running 2.0, I suppose it would be easier to upgrade to 2.1. In our case it's the whole OS upgrade.

**Comment by TimWolla** [comment]
> Please do me a favor, have a look around to see if any distro already ships 2.0, that will help me :-)

Ubuntu Eaon (19.10) will ship with HAProxy 2.0.5 on October, 17th. It's not a LTS Ubuntu, though. So hopefully no one will run that on servers.

see: https://packages.ubuntu.com/eoan/haproxy and https://www.omgubuntu.co.uk/2019/05/ubuntu-19-10-release-features

**Comment by wtarreau** [comment]
Well, first, we never ever backport features to older versions. It may occasionally happen for the latest LTS release if really needed but that's all. In addition backporting CONTINUATION to 1.8 doesn't make sense since it's not even supported in the parser! Also there's no relation between your version and the OS version : either you're running a version from sources that you build yourself and you will not depend on the OS, or you're using the version shipped with the OS and you will never get any such backport anyway (nor most fixes by the way).

The last (and recommended option) when using debian/ubuntu is to use the bpackages from haproxy.debian.net that are kept up to date for your distro. However there is no jessie support there for 2.0. In your case I would suggest that you simply compile it yourself, it has very little dependencies and I don't see any reason for you to run into trouble.

**Comment by wtarreau** [comment]
Hi Tim,

OK good to know, thanks! Anyway, as usual with debian/ubuntu, these versions will miss the vast majority of fixes for all subsequent bugs and users will all be strongly encouraged to use Vincent's packages instead, so that's not a big deal. I.e. if someone says "I'm running 2.0.5 on ubuntu", the first response he'll get will be "outdated 200 known bugs, please update". It's harder when it happens with distros which do not have a serious parallel channel for releases (and even worse when they are LTS).

**Comment by wtarreau** [comment]
Implemented and merged into 2.1 more than a year ago, closing now.
